Assets/_BForBoss/_Utility/Editor/PropertyDrawers/ResolverPropertyDrawer.cs

```diff
@@ -0,0 +1,369 @@
+
+            Type fieldType = fieldInfo.FieldType;
+            _isArray = fieldType.IsArray;
+            _isList = fieldType.IsGenericType && fieldType.GetGenericTypeDefinition() == typeof(List<>);
+            
+            //Check if field is being declared within Struct
+            Type declaringType = fieldInfo.DeclaringType;
+            //https://stackoverflow.com/questions/1827425/how-to-check-programmatically-if-a-type-is-a-struct-or-a-class
+            _isDeclaredInStruct = declaringType != null && declaringType.IsValueType && !declaringType.IsPrimitive &&
+                                  declaringType != typeof(decimal) && declaringType != typeof(DateTime) &&
+                                  !declaringType.IsEnum;
+
+            position.Set(position.x, position.y, position.width - RESOLVER_WIDTH, position.height);
+            EditorGUI.PropertyField(position,property,label);
+
+            CreateRects(position);
+
+            DrawResolverButtons(property);
+        }
+
+        private void DrawResolverButtons(SerializedProperty property)
+        {
+            // Self (G) | Children (C) | Parent (P) | (OPTIONAL) Sibilings (R) | None (X)
+            
+            if (GUI.Button(_childResolveRect, _childResolveContent))
+            {
+                SetResolvedComponent(property, ResolveType.FromChildren);
+            }
+            
+            if (GUI.Button(_parentResolveRect, _parentResolveContent))
+            {
+                SetResolvedComponent(property, ResolveType.FromParent);
+            }
+            
+            if (GUI.Button(_selfResolveRect, _selfResolveContent))
+            {
+                SetResolvedComponent(property, ResolveType.FromSelf);
+            }
+            
+            if (GUI.Button(_clearContentRect, _clearContent, _clearContentGUIStyle))
+            {
+                Undo.RegisterCompleteObjectUndo(property.serializedObject.targetObject, "Cleared Resolved Component");
+                property.objectReferenceValue = null;
+            }
+        }
+
+        private void SetResolvedComponent(SerializedProperty property, ResolveType resolveType)
+        {
+            Undo.RegisterCompleteObjectUndo(property.serializedObject.targetObject, "Auto Resolved Component");
+            MonoBehaviour mb = property.serializedObject.targetObject as MonoBehaviour;
+            Type fieldType = _isArray ? fieldInfo.FieldType.GetElementType() :
+                _isList ? fieldInfo.FieldType.GetGenericArguments().Single() : fieldInfo.FieldType;
+
+            bool isGameObjectType = fieldType == typeof(GameObject);
+
+            List<GameObject> gameObjects = new List<GameObject>();
+            Component[] components = null;
+
+            switch (resolveType)
+            {
+                case ResolveType.FromChildren:
+                {
+                    if (isGameObjectType)
```

Comment:
In terms of cleaning up, you can always abstract this out into separate methods so it's more readable

Reply:
Yeah in reality, both `Components` and `GameObjects` derive from `Object` so I shouldn't really need two separate implementations. And instead of using `GetInstanceID()` to get the ID for ObjectReference I could use `GetHashCode` which Object class implements.

Could possibly fix the `ScriptableObject` derivation issue as well now that I think about it